docs/puzzles/day21.md

```diff
@@ -1,11 +1,312 @@
+
+case class Player(cell: Cell, score: Long)
+
+type Players = (Player, Player)
+
+final class DeterministicDie {
+  var throwCount: Int = 0
+  private var lastValue: Int = 100
+
+  def nextResult(): Int =
+    throwCount += 1
+    lastValue = (lastValue % 100) + 1
+    lastValue
+}
+
+def part1(input: String): Long =
+  val players = parseInput(input)
+  val die = new DeterministicDie
+  val loserScore = playWithDeterministicDie(players, die)
+  loserScore * die.throwCount
+
+def parseInput(input: String): Players =
+  val lines = input.split("\n")
+  (parsePlayer(lines(0)), parsePlayer(lines(1)))
+
+def parsePlayer(line: String): Player =
+  line match
+    case s"Player $num starting position: $cell" =>
+      Player(cell.toInt - 1, 0L)
+
+@tailrec
+def playWithDeterministicDie(players: Players, die: DeterministicDie): Long =
+  val diesValue = die.nextResult() + die.nextResult() + die.nextResult()
+  val player = players(0)
+  val newCell = (player.cell + diesValue) % 10
+  val newScore = player.score + (newCell + 1)
+  if newScore >= 1000 then
+    players(1).score
+  else
+    val newPlayer = Player(newCell, newScore)
+    playWithDeterministicDie((players(1), newPlayer), die)
+```
+
+<Solver puzzle="day21-part1"/>
+
+## The Dirac die
+
+For part 2, our previous model falls short.
+We now have to simulate a large number of universes.
+
+The good thing is that our die is not stateful anymore, and we do not need to keep track of how many times it is thrown, so we do not use an instance for it at all.
+
+However, we will have to count how many times each player wins, which I chose to keep track using mutable state again:
+
+```scala
+final class Wins(var player1Wins: Long, var player2Wins: Long)
+
+def part2(input: String): Long =
+  val players = parseInput(input)
+  val wins = new Wins(0L, 0L)
+  // ... Simulate universes here
+  Math.max(wins.player1Wins, wins.player2Wins)
+```
+
+## Naive solution (non-practical)
+
+A first attempt would be to use a (non-tail) recursive function that branches out for every possible outcome of the three dice.
+That would look like the following:
+
+```scala
+def playWithDiracDie(players: Players, player1Turn: Boolean, wins: Wins): Unit =
+  for
+    die1 <- List(1, 2, 3)
+    die2 <- List(1, 2, 3)
+    die3 <- List(1, 2, 3)
+  do
+    val diesValue = die1 + die2 + die3
+    val player = players(0)
+    val newCell = (player.cell + diesValue) % 10
+    val newScore = player.score + (newCell + 1)
+    if newScore >= 21 then
+      if player1Turn then
+        wins.player1Wins += 1L
+      else
+        wins.player2Wins += 1L
+    else
+      val newPlayer = Player(newCell, newScore)
+      playWithDiracDie((players(1), newPlayer), !player1Turn, wins)
+  end for
+```
+
+For every possible outcome, we compute the new cell and score.
+If the game is over, we add 1 to the number of ties that the current player wins.
+Otherwise, we recursive for the next roll.
```

Comment:
Is it correct to say “we recursive for”?

Reply:
Nope. It's "we recurse", indeed. ;)